src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/OpenSslX509ChainEventSource.cs

```diff
@@ -0,0 +1,728 @@
+        private const int EventId_FlushStores = 3;
+        private const int EventId_FindFirstChainFinished = 4;
+        private const int EventId_FindChainViaAiaFinished = 5;
+        private const int EventId_AiaDisabled = 6;
+        private const int EventId_NoAiaFound = 7;
+        private const int EventId_InvalidAia = 8;
+        private const int EventId_NonHttpAiaEntry = 9;
+        private const int EventId_AssetDownloadStart = 10;
+        private const int EventId_AssetDownloadStop = 11;
+        private const int EventId_HttpClientNotAvailable = 12;
+        private const int EventId_DownloadTimeExceeded = 13;
+        private const int EventId_InvalidDownloadedCertificate = 14;
+        private const int EventId_InvalidDownloadedCrl = 15;
+        private const int EventId_InvalidDownloadedOcsp = 16;
+        private const int EventId_DownloadCompleteStatusCode = 17;
+        private const int EventId_DownloadRedirected = 18;
+        private const int EventId_DownloadRedirectsExceeded = 19;
+        private const int EventId_DownloadRedirectNotFollowed = 20;
+        private const int EventId_UntrustedChainWithRevocation = 21;
+        private const int EventId_NonHttpCdpEntry = 22;
+        private const int EventId_NoCdpFound = 23;
+        private const int EventId_NoMatchingAiaEntry = 24;
+        private const int EventId_NoMatchingCdpEntry = 25;
+        private const int EventId_CrlCacheCheckStart = 26;
+        private const int EventId_CrlCacheCheckStop = 27;
+        private const int EventId_CrlCacheOpenError = 28;
+        private const int EventId_CrlCacheDecodeError = 29;
+        private const int EventId_CrlCacheExpired = 30;
+        private const int EventId_CrlCacheFileBasedExpiry = 31;
+        private const int EventId_CrlCacheAcceptedFile = 32;
+        private const int EventId_CrlCacheWriteFailed = 33;
+        private const int EventId_CrlCacheWriteSucceeded = 34;
+        private const int EventId_CrlCheckOffline = 35;
+        private const int EventId_CrlChainFinished = 36;
+        private const int EventId_AllRevocationErrorsIgnored = 37;
+        private const int EventId_OcspResponseFromCache = 38;
+        private const int EventId_OcspResponseFromDownload = 39;
+        private const int EventId_RawElementStatus = 40;
+        private const int EventId_FinalElementStatus = 41;
+        private const int EventId_CouldNotOpenCAStore = 42;
+        private const int EventId_CachingIntermediate = 43;
+        private const int EventId_CachingIntermediateFailed = 44;
+        private const int EventId_RevocationCheckStart = 45;
+        private const int EventId_RevocationCheckStop = 46;
+        private const int EventId_CrlIdentifiersDetermined = 47;
+
+        private static string GetCertificateSubject(SafeX509Handle certHandle)
+        {
+            X500DistinguishedName dn = Interop.Crypto.LoadX500Name(Interop.Crypto.X509GetSubjectName(certHandle));
```

Comment:
Is this thread safe if the `certHandle` is disposed? `X509GetSubjectName` returns a raw pointer, `certHandle` could get disposed between `LoadX500Name` and `X509GetSubjectName`.

I'm basically looking to avoid a repeat of https://github.com/dotnet/runtime/issues/49732. If the race can't happen, cool. Otherwise maybe we need to do a `DangerousAddRef` somewhere.

Reply:
In the cases that it gets used (during chain building) it isn't susceptible to being freed because its lifetime is tied back to the `X509_STORE_CTX*`, which is in a using-dispose for the whole of the chain build.

We could harden it though, just in case we use it in a weird place later.

Reply:
> We could harden it though, just in case we use it in a weird place later.

Up to you. On one hand, as you say, its current use doesn't do the bad behavior and it's `private`. On the other hand, incorrect use amounts to a UAF.